### PR TITLE
Trim the subject before encrypting the subject

### LINKED ISSUE #346: NC16.0.1 - Notification (app 2.4.1) broken
### Steps to reproduce
Posting an announcement via the announcement center to any group. 

### Expected behaviour
a red dot warning should appear in the header on top_right page ( notification API i believe )

### Actual behaviour
The announcement shows up un the announcement center page = correct.
The announcement shows up un the header_right (class notification) icon = BROKEN

## Server configuration detail

**Operating system:** Debian 10.0 X86_64

**Webserver:** Apache (fpm-fcgi)

**Database:** mysql 10.1.38

**PHP version:** 

7.3.5
Modules loaded:
Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, sodium, session, standard, cgi-fcgi, mysqlnd, PDO, xml, apcu, bcmath, bz2, calendar, ctype, curl, dom, mbstring, fileinfo, ftp, gd, gettext, iconv, igbinary, imagick, intl, json, ldap, exif, mysqli, pdo_mysql, apc, posix, readline, redis, shmop, SimpleXML, smbclient, soap, sockets, sysvmsg, sysvsem, sysvshm, tokenizer, wddx, xmlreader, xmlwriter, xsl, zip, Phar, libsmbclient, Zend OPcache

**Nextcloud version:** 16.0.1 - 16.0.1.1

**Updated from an older Nextcloud/ownCloud or fresh install:** NC13/NC14/NC15

<details><summary>Dropdown List of activated apps</summary>

```
Enabled:
 - accessibility: 1.2.0
 - activity: 2.9.1
 - admin_audit: 1.6.0
 - announcementcenter: 3.5.1
 - apporder: 0.7.1
 - audioplayer: 2.7.0
 - browser_warning: 0.1.0
 - bruteforcesettings: 1.3.0
 - calendar: 1.7.0
 - camerarawpreviews: 0.7.0
 - checksum: 0.4.2
 - cloud_federation_api: 0.2.0
 - comments: 1.6.0
 - contacts: 3.1.1
 - data_request: 1.3.0
 - dav: 1.9.2
 - external: 3.3.0
 - federatedfilesharing: 1.6.0
 - federation: 1.6.0
 - files: 1.11.0
 - files_antivirus: 2.1.1
 - files_downloadactivity: 1.5.0
 - files_pdfviewer: 1.5.0
 - files_rightclick: 0.13.0
 - files_sharing: 1.8.0
 - files_texteditor: 2.8.0
 - files_trackdownloads: 1.5.0
 - files_trashbin: 1.6.0
 - files_versions: 1.9.0
 - files_videoplayer: 1.5.0
 - firstrunwizard: 2.5.0
 - flowupload: 0.1.0
 - gallery: 18.3.0
 - groupfolders: 4.0.2
 - impersonate: 1.3.0
 - issuetemplate: 0.5.0
 - logreader: 2.1.0
 - lookup_server_connector: 1.4.0
 - metadata: 0.9.0
 - notifications: 2.4.1
 - oauth2: 1.4.2
 - occweb: 0.0.4
 - password_policy: 1.6.0
 - previewgenerator: 2.1.0
 - privacy: 1.0.0
 - provisioning_api: 1.6.0
 - quota_warning: 1.5.0
 - recommendations: 0.4.0
 - registration: 0.4.6
 - serverinfo: 1.6.0
 - sharebymail: 1.6.0
 - sharerenamer: 2.6.0
 - socialsharing_facebook: 1.0.4
 - support: 1.0.0
 - survey_client: 1.4.0
 - systemtags: 1.6.0
 - twofactor_backupcodes: 1.5.0
 - twofactor_totp: 2.1.2
 - twofactor_u2f: 3.0.0
 - twofactor_yubikey: 0.4.1
 - unsplash: 1.1.3
 - updatenotification: 1.6.0
 - user_usage_report: 1.1.1
 - viewer: 1.0.0
 - workflowengine: 1.6.0
Disabled:
 - circles
 - dashboard
 - encryption
 - files_external
 - music
 - news
 - nextcloud_announcements
 - polls
 - sociallogin
 - theming
 - user_ldap
 - user_saml

```
</details>

<details><summary>Dropdown Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "nextcloud.rkn.ovh"
    ],
    "memcache.distributed": "\\OC\\Memcache\\Redis",
    "memcache.locking": "\\OC\\Memcache\\Redis",
    "memcache.local": "\\OC\\Memcache\\APCu",
    "redis": {
        "host": "***REMOVED SENSITIVE VALUE***",
        "port": "***REMOVED SENSITIVE VALUE***",
    },
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "overwrite.cli.url": "https:\/\/nextcloud.rkn.ovh",
    "dbtype": "mysql",
    "version": "16.0.1.1",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "mail_smtpmode": "smtp",
    "mail_smtpauthtype": "LOGIN",
    "mail_smtpsecure": "ssl",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpauth": 1,
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "cache_chunk_gc_ttl": 86400,
    "filelocking.ttl": 3600,
    "auth.bruteforce.protection.enabled": false,
    "maintenance": false,
    "loglevel": 1,
    "ldapIgnoreNamingRules": false,
    "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
    "theme": "",
    "updater.release.channel": "stable",
    "twofactor_enforced": "true",
    "twofactor_enforced_groups": [],
    "twofactor_enforced_excluded_groups": [],
    "mysql.utf8mb4": true,
    "app_install_overwrite": [
        "external",
        "occweb",
        "files_downloadactivity",
        "admin_notifications",
        "twofactor_yubikey"
    ]
}
```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** "***REMOVED SENSITIVE VALUE***",

**Are you using an external user-backend, if yes which one:** no

## Logs

<details><summary>Web server error log</summary>

```
no logs
```
</details>

<details><summary>Nextcloud log</summary>

```
# Logs was empty before testing
"Could not find resource js\/oc-backbone-webdav.js to load"
"Could not find resource vendor\/select2\/select2.js to load"
"Could not find resource vendor\/select2\/select2.css to load"
```
</details>

<details><summary>Browser log</summary>

```
no logs
```

</details>

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #346

### CODE REVIEW

File: lib/Push.php [hunk 1]
```diff
@@ -187,11 +187,18 @@ protected function encryptAndSign(Key $userKey, array $device, int $id, INotific
...
+			'subject' => '',
 			'type' => $notification->getObjectType(),
 			'id' => $notification->getObjectId(),
 		];
 
+		// Max length of encryption is 255, so we need to shorten the subject to be shorter
+		$subject = $notification->getParsedSubject();
+		$dataLength = 245 - strlen(json_encode($data));
+		if (strlen($subject) > $dataLength) {
+			$data['subject'] = substr($subject, 0, $dataLength) . '…';
```
Comment (rullzer) [inline lib/Push.php]: Should the substr not be 1 shorter than the dataLength?

Comment (nickvergessen) [reply lib/Push.php]: I started with 245, instead of 255 so we have a little buffer there

Comment (rullzer) [reply lib/Push.php]: Ah ok perfect. I wasn't sure if that was maybe for other metadata somewhere.

**Pull request merged**